internal/policy/verify.go

```diff
@@ -83,6 +83,186 @@ func VerifyRefFull(ctx context.Context, repo *gitinterface.Repository, target st
+			verifier := verifier
+			verifier.threshold--
+
+			err := verifier.Verify(ctx, gitinterface.ZeroHash, authorization, githubApproval)
+			if err == nil {
+				rslEntrySignatureNeededForThreshold = true
+				gitNamespaceVerified = true
+				break
+			} else if !errors.Is(err, ErrVerifierConditionsUnmet) {
+				return false, err
+			}
+		}
+	}
+
+	if !gitNamespaceVerified {
+		return false, fmt.Errorf("not enough approvals to meet Git namespace policies, %w", ErrUnauthorizedSignature)
+	}
+
+	hasFileRule, err := policyState.hasFileRule()
+	if err != nil {
+		return false, err
+	}
+	if !hasFileRule {
+		return rslEntrySignatureNeededForThreshold, nil
+	}
+
+	// Verify modified files
+	commitIDs, err := repo.GetCommitsBetweenRange(featureEntry.TargetID, targetEntry.TargetID)
+	if err != nil {
+		return false, err
+	}
+
+	commitsVerified := make([]bool, len(commitIDs))
+	for i, commitID := range commitIDs {
+		// Assume the commit's paths are verified, if a path is left unverified,
+		// we flip this later.
+		commitsVerified[i] = true
+
+		paths, err := repo.GetFilePathsChangedByCommit(commitID)
+		if err != nil {
+			return false, err
+		}
+
+		pathsVerified := make([]bool, len(paths))
+		verifiedUsing := ""
+		for j, path := range paths {
+			verifiers, err := policyState.FindVerifiersForPath(fmt.Sprintf("%s:%s", fileRuleScheme, path))
+			if err != nil {
+				return false, err
+			}
+
+			if len(verifiers) == 0 {
+				pathsVerified[j] = true
+				continue
+			}
+
+			if len(verifiedUsing) > 0 {
+				// We've already verified and identified commit signature, we
+				// can just check if that verifier is trusted for the new path.
+				// If not found, we don't make any assumptions about it being a
+				// failure in case of name mismatches. So, the signature check
+				// proceeds as usual.
+				for _, verifier := range verifiers {
+					if verifier.Name() == verifiedUsing {
+						pathsVerified[j] = true
+						break
+					}
+				}
+			}
+
+			if pathsVerified[j] {
+				continue
+			}
+
+			for _, verifier := range verifiers {
+				err := verifier.Verify(ctx, commitID, authorization, githubApproval)
+				if err == nil {
+					// Signature verification succeeded
+					pathsVerified[j] = true
+					verifiedUsing = verifier.Name()
+					break
+				} else if !errors.Is(err, ErrVerifierConditionsUnmet) {
+					// Unexpected error
+					return false, err
+				}
+			}
+		}
+
+		for _, p := range pathsVerified {
+			if !p {
+				// Flip earlier assumption that commit paths are verified as we
+				// find that at least one path wasn't verified successfully
+				commitsVerified[i] = false
+				break
+			}
+		}
+	}
+
+	for _, c := range commitsVerified {
+		if !c {
+			pathNamespaceVerified = false
+			break
+		}
+	}
+
+	if !pathNamespaceVerified {
+		return false, fmt.Errorf("not enough approvals to meet file namespace policies, %w", ErrUnauthorizedSignature)
+	}
+
+	return rslEntrySignatureNeededForThreshold, nil
```

Comment:
Taking a step back, there's a lot of verification logic here that feels like should be encapsulated elsewhere. Ideally I think this should look like:

What information do you need to be returned from verification in order to make policy decisions about mergability?